### ISSUE TITLE
ParallelTask window_id actually chunk_id?

### ISSUE THREAD

**Issue by mmann1123** [jgrss/geowombat#57]
Ok after a lot of fumbling around, I think I figured out an issue I have been having. It seems like in `data, window_id, num_workers = list(itertools.chain(*args)) ` , the `window_id` is actually the chunk_id. I came across this because of the unusual size of my data, but managed to finally figure out a way to replicate it. 

If n_windows is ever larger than n_chunks, the window_id iterates through values of n_chunks, but then repeats itself until you reach window_id. So in my case I keep overwriting my file.

Here's an example 

``` python 
import numpy as np
import dask.array as da
import xarray as xr
from geowombat.core.parallel import ParallelTask
import geowombat as gw
import itertools
import os 
sys.path.append('/home/mmann1123/Documents/github/xr_fresh/')
from xr_fresh.utils import save_pickle,open_pickle, unique,add_categorical,add_time_targets,compressed_pickle,decompress_pickle
from xr_fresh.transformers import Stackerizer
from geowombat.data import rgbn_20160101, rgbn_20160401, rgbn_20160517


out_folder =   '~/Desktop/'
 
def user_func(*args):

    data, window_id, num_workers = list(itertools.chain(*args))    # num_workers=1
    X = Stackerizer(stack_dims = ('y','x','time'), direction='stack').fit_transform(data)   # NOTE stack y before x!!! 
    X = X.to_pandas() 
    X.columns = X.columns.astype(str)

    X.to_parquet(out_folder+'/testing/test_'+  '_%05d.parquet' %  window_id)

    return None


multiplier = 50

import geowombat as gw
with gw.open([rgbn_20160101, rgbn_20160401, rgbn_20160517]*multiplier,
                 band_names=['blue', 'green', 'red','nir'],
                 time_names=[str(x) for x in range(multiplier*3)],chunks=25) as src:
        print(src)

        print('stack:')

        pt = ParallelTask(src,
                    scheduler='threads',
                    n_workers=3)
        print('n chunks %s' % pt.n_chunks)
        print('n windows %s' % pt.n_windows)
        res = pt.map(user_func, 4) 
```

**Comment by mmann1123** [comment]
A a simple solution we could add a chunk id in the generator as follows:

``` python  
class ParallelTask

def map(self, func, *args):
    ....
    for cidx, wchunk in enumerate(range(0, self.n_windows, self.n_chunks)):

                data_gen = ((self.data[slice_], cidx, widx, *args) for widx, slice_ in enumerate(window_slice))

```
although I am admittedly a bit confused about chunk vs  window, especially as this relates to pt.n_chunks pt.n_windows

**Comment by mmann1123** [comment]
Found a better option (see PR), the following will ensure that the window_id increments from 0 the n_windows without repeat. 

``` python
class ParallelTask

def map(self, func, *args):
    ....
    for cidx, wchunk in enumerate(range(0, self.n_windows, self.n_chunks)):

                data_gen = ((self.data[slice_], cidx*self.n_chunks + widx, *args) for widx, slice_ in enumerate(window_slice))

```

**Comment by jgrss** [comment]
Nice catch. I should have used `batches` instead of `chunks`. You can think of them as mini batches of windows passed to the parallel executor. I added it a while back to limit memory overhead.

I think we just need the following, right?

```python
for wchunk in range(0, self.n_windows, self.n_chunks):

    data_gen = ((self.data[slice_], widx+wchunk, *args) for widx, slice_ in enumerate(window_slice))
```

**Comment by mmann1123** [comment]
I think you need to multiply to avoid overwriting. Play around with my
example.

**Comment by jgrss** [comment]
@mmann1123 The window id was also incorrect in `to_raster`. Both `cidx*self.n_chunks + widx` and `widx+wchunk` produced the same result. The fix was pushed in [911145ecc92bad22f06d0326cadf99d46ec893f2](https://github.com/jgrss/geowombat/commit/911145ecc92bad22f06d0326cadf99d46ec893f2).

**Comment by mmann1123** [comment]
Ok great. Can we close this?
